Use etcd3 for both the temporary and self-hosted control planes

Quentin-M opened the pull request:
The existing `--storage-backend=etcd3` parameter can lead to unexpected
and confusing behavior when used without the experimental self-hosted
etcd (`--experimental-self-hosted-etcd=true`) or with the examples.

When specified, Bootkube runs the temporary control plane against etcd
v2 and then the self-hosted plane using etcd v3. However, etcd uses two
separate data stores for v2 and v3 APIs (https://git.io/v1NxV).
Therefore when the transition occurs after the self-hosted control plane
has been created, everything basically disappears.

Instead of adding yet another flag to Bootkube to enable etcd3 with the
temporary control-plane, this commit makes a leap forward and use etcd3
as the default storage backend for both control-planes, which will be the 
default in Kubernetes 1.6.

**Note: This PR shouldn't be merged before CoreOS Linux Container 1214+ is promoted Stable, on Jan 3, 2017.**

Quentin-M commented:
Jenkins is broken:

```
[bootkube-conformance] $ /bin/sh -xe /tmp/hudson4072966781119609134.sh
+ set +x
docker: "login" requires a maximum of 1 argument.
See 'docker login --help'.

Usage:	docker login [OPTIONS] [SERVER]

Log in to a Docker registry.
If no server is specified, the default is defined by the daemon.
Build step 'Execute shell' marked build as failure
Setting status of 669a64b507538c7550f1d1cc1510e7dad836fcbe to FAILURE with url https://jenkins-tectonic.prod.coreos.systems/job/bootkube-conformance/5/ and message: 'Build finished. '
Using context: bootkube-dev
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
Finished: FAILURE
```

aaronlevy commented:
@pbx0 would you be able to take a look at the jenkins builder?

aaronlevy commented:
@xiang90 or @hongchaodeng mind doing a quick sanity check RE the etcd changes

xiang90 commented:
@Quentin-M Basically we change everything to run etcd3. no etcd2 will ever get involved?

Quentin-M commented:
@xiang90 Yes.

xiang90 commented:
@Quentin-M the change looks good to me.

@hongchaodeng can you take a look?

Quentin-M commented:
hack/single-node: OK
hack/multi-node: OK

aaronlevy commented:
Closing and re-opening to force rerun of CI.
@pbx0 is there a comment I can use to rerun instead?

aaronlevy commented:
Also just noting that merging this may be blocked until the [etcd-member.service](https://github.com/coreos/coreos-overlay/blob/master/app-admin/etcd-wrapper/files/etcd-member.service) is available in a stable release (>= 1214.0.0).

pbx0 commented:
@aaron I left triggers for ghprb jenkins plugin default for bootkube dev. So options are:
```
"ok to test" to accept this pull request for testing
"test this please" for a one time test run
"add to whitelist" to add the author to the whitelist
If the build fails for other various reasons you can rebuild.

"retest this please" to start a new build
```
I can change the trigger to be more like checkpointer-dev but then they would conflict I think.

aaronlevy commented:
nope that's perfect, thanks

Quentin-M commented:
test this please

Quentin-M commented:
Conformance tests can't work due to the fact that it still uses etcd v2.
Created PR on Pluton: https://github.com/coreos-inc/pluton/pull/16

Quentin-M commented:
retest this please

pbx0 commented:
retest this please

Quentin-M commented:
Can't run conformance test: Jenkins is broken.

Quentin-M commented:
test this please

Quentin-M commented:
test this please

Quentin-M commented:
```
Ran 117 of 473 Specs in 6977.489 seconds
109 Passed | 8 Failed | 0 Pending | 356 Skipped --- FAIL: TestE2E (6977.51s)

Summarizing 8 Failures:
[Fail] [k8s.io] Pods [It] should be submitted and removed [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:250
[Fail] [k8s.io] Pods Delete Grace Period [It] should be submitted and removed [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:176
[Fail] [k8s.io] Downward API volume [It] should provide container's cpu limit [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2167
[Fail] [k8s.io] Downward API volume [It] should provide container's cpu request [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2167
[Fail] [k8s.io] Downward API volume [It] should provide container's memory limit [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2167
[Fail] [k8s.io] Downward API [It] should provide pod IP as an env var [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2167
[Fail] [k8s.io] Downward API [It] should provide pod name and namespace as env vars [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2167
[Fail] [k8s.io] Kubectl client [k8s.io] Guestbook application [It] should create and stop a working application [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1577
```

Quentin-M commented:
E2E tests passed on upstream Kubernetes with etcd3: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39291/pull-kubernetes-e2e-gce-etcd3/10952/

Quentin-M commented:
[CoreOS 1235.4.0](https://coreos.com/releases/#1235.4.0) has been released.
[Self-hosted flannel](https://github.com/kubernetes-incubator/bootkube/pull/247) has been merged.

PTAL.

@aaronlevy

Quentin-M commented:
```
Ran 117 of 473 Specs in 5679.515 seconds
FAIL! -- 113 Passed | 4 Failed | 0 Pending | 356 Skipped --- FAIL: TestE2E (5679.54s)

Summarizing 4 Failures:
[Fail] [k8s.io] Pods Delete Grace Period [It] should be submitted and removed [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:176
[Fail] [k8s.io] Networking [k8s.io] Granular Checks: Pods [It] should function for node-pod communication: http [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/networking_utils.go:544
[Fail] [k8s.io] Pods [It] should be submitted and removed [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:250
[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates resource limits of pods that are allowed to run [Conformance] 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:200
```

aaronlevy commented:
We need to determine why we're getting those failures before this can merge (or if you run the tests against a normal baseline and we get same errors, then it wasn't introduced here).

Quentin-M commented:
All the tests failed because the scheduler refusing the create the pods due to insufficient CPU. Looking to spawn bigger machines and re-test.

Quentin-M commented:
```
Ran 117 of 473 Specs in 4789.253 seconds
FAIL! -- 114 Passed | 3 Failed | 0 Pending | 356 Skipped --- FAIL: TestE2E (4789.27s)

Summarizing 3 Failures:
[Fail] [k8s.io] Pods Delete Grace Period [It] should be submitted and removed [Conformance] 
[Fail] [k8s.io] Pods [It] should be submitted and removed [Conformance] 
[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates resource limits of pods that are allowed to run [Conformance] 
```

pbx0 commented:
pluton smoke and destruction tests pass. coreosbot is temporarily offline.

Quentin-M commented:
@xiang90 @hongchaodeng Do you mind taking a look at this and telling us if the three remaining issues are relevant? The full logs are available on Build #89.

pbx0 commented:
Ive seen `[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates resource limits of pods that are allowed to run [Conformance] ` failures many many times before so I don't think that failure is introduced here. Not sure about the other two.

Quentin-M commented:
# Fixed all E2E failures

## Explanation
As of CoreOS Container Linux 1298.1.0, `etcd-member.service` [spawns](https://github.com/coreos/coreos-overlay/blob/85f3b0a/app-admin/etcd-wrapper/files/etcd-member.service#L15) etcd v3.0.10 by default. However, Kubernetes`s cacher requires the etcd server to return the previous key-value (`prevValue`), as the logs below suggest. The first etcd version to do that is [v3.0.11](https://github.com/coreos/etcd/blob/master/NEWS#L44
). Solving the issue boiled down to making `etcd-member.service` start etcd v3.1.0.

```
W0127 19:17:08.045538       7 listers.go:69] can not retrieve list of objects using index : Index with name namespace does not exist
E0127 19:17:11.359210       7 reflector.go:359] pkg/storage/cacher.go:201: expected type *api.Pod, but watch event object had type <nil>
```

## E2E reports from previously failing tests

```
• [SLOW TEST:16.799 seconds]
[k8s.io] Pods Delete Grace Period
/go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:826
  should be submitted and removed [Conformance]
  /go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:188
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

• [SLOW TEST:15.919 seconds]
[k8s.io] Pods
/go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:826
  should be submitted and removed [Conformance]
  /go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:261
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Ran 2 of 473 Specs in 32.792 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 471 Skipped PASS

Ginkgo ran 1 suite in 33.090621861s
Test Suite Passed
2017/01/27 19:31:31 e2e.go:724: Step 'Ginkgo tests' finished in 33.411707635s
```

```
• [SLOW TEST:105.292 seconds]
[k8s.io] SchedulerPredicates [Serial]
/go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:826
  validates resource limits of pods that are allowed to run [Conformance]
  /go/src/github.com/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:214
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 473 Specs in 105.363 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 472 Skipped PASS
```

Now re-running the E2E suite with the missing version bumps: https://github.com/kubernetes-incubator/bootkube/pull/291 https://github.com/kubernetes-incubator/bootkube/pull/290

Quentin-M commented:
```
Ran 117 of 473 Specs in 2643.345 seconds
SUCCESS! -- 117 Passed | 0 Failed | 0 Pending | 356 Skipped PASS

Ginkgo ran 1 suite in 44m3.617445336s
```

Quentin-M commented:
PTAL.
